### PR TITLE
.insn directive support

### PR DESCRIPTION
**Pull request by kito-cheng**
Implement .insn directive

https://github.com/riscv/riscv-binutils-gdb/issues/38

- [x] .insn support for assembler 
- [x] Add testcase for .insn
- [x] Document for .insn
- [x] Add symbolic name for opcode
 - eg. Able to use `LOAD`, `custom_1` for opcode field

### CODE REVIEW

**Comment by kito-cheng** [comment]
Do NOT merge yet, document and testing not done.
Just send PR to prevent other guys do the same work.

**Comment by sorear** [comment]
Thank you for picking this up

File: opcodes/riscv-opc.c [hunk 1]
```diff
@@ -686,3 +686,29 @@ const struct riscv_opcode riscv_opcodes[] =
...
+const struct riscv_opcode riscv_insn_types[] =
+{
+/* name,  isa,           operands, match, mask,    match_func, pinfo.  */
+{"r",     "I",   "O4,F3,F7,d,s,t",     0,    0,  match_opcode, 0 },
+{"i",     "I",   "O4,F3,F7,d,s,j",     0,    0,  match_opcode, 0 },
+{"s",     "I",   "O4,F3,F7,t,s,q",     0,    0,  match_opcode, 0 },
+{"s",     "I",  "O4,F3,F7,t,q(s)",     0,    0,  match_opcode, 0 },
+{"sb",    "I",  "O4,F3,F7,d,o(s)",     0,    0,  match_opcode, 0 },
+{"u",     "I",     "O4,F3,F7,d,u",     0,    0,  match_opcode, 0 },
+{"uj",    "I",     "O4,F3,F7,d,a",     0,    0,  match_opcode, 0 },
```
Comment (aswaterman) [inline opcodes/riscv-opc.c]: ```sb``` and ```uj``` should be renamed ```b``` and ```j``` (the names changed in v2.2 of the spec).

**Comment by aswaterman** [comment]
Cool!

**Comment by palmer-dabbelt** [comment]
Thanks a bunch, @kito-cheng.  When you're ready just ping me.

File: opcodes/riscv-opc.c (hunk 1 above)
Comment (kito-cheng) [reply opcodes/riscv-opc.c]: Updated

**Comment by aswaterman** [comment]
@kito-cheng any word on testing & documentation?

**Comment by kito-cheng** [comment]
I expect it can finish in my Chinese new year vocation :P 
I've rebase to 2.30 and add several testcase now.

**Comment by kito-cheng** [comment]
I've written document and testcase, but I think it's time to rebase riscv-next branch?

**Comment by kito-cheng** [comment]
This PR already merged in upstream. 
https://sourceware.org/git/gitweb.cgi?p=binutils-gdb.git;a=commit;h=0e35537d754f1c687850d1caccb2d78d2e418391

**Pull request closed without merging**